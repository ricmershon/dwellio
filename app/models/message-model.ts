import { Document, Schema, Types, model, models } from 'mongoose';
import { MessageInputType } from '@/app/schemas/message-schema';

export interface MessageDocument extends MessageInputType, Document {
    sender: Types.ObjectId,
    recipient: Types.ObjectId,
    property: Types.ObjectId | { _id: Types.ObjectId, name: string },
    read: boolean,
    createdAt: Date;
    updatedAt: Date;
}

const MessageSchema = new Schema({
    sender: { type: Schema.Types.ObjectId, ref: 'User', required: true },
    recipient: { type: Schema.Types.ObjectId, ref: 'User', required: true },
    property: { type: Schema.Types.ObjectId, ref: 'Property', required: true },
    name: { type: String, required: [true, 'Name is required'] },
    email: { type: String, required: [true, 'Email is required'] },
    phone: String,
    body: String,
    read: { type: Boolean, default: false }
}, {
    timestamps: true
});

const Message = models.Message || model<MessageDocument>('Message', MessageSchema);
export default Message;