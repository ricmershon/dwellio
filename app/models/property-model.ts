import { Document, Schema, Types, model, models } from 'mongoose';

import { PropertyImageData } from '@/app/lib/definitions';
import { PropertyInputType } from '@/app/schemas/property-schema';

export interface PropertyDocument extends Omit<PropertyInputType, 'imagesData'>, Document {
    owner: Types.ObjectId;
    isFeatured?: boolean;
    imagesData?: PropertyImageData[];
    createdAt: Date;
    updatedAt: Date;
}

const ImageSchema = new Schema({
    secureUrl: { type: String, required: true },
    publicId: { type: String, required: true }
}, { _id: false });

const PropertySchema = new Schema<PropertyDocument>({
    owner: { type: Schema.Types.ObjectId, required: true, ref: 'User' },
    name: { type: String, required: true },
    type: { type: String, required: true },
    description: { type: String  },
    location: {
        street: String,
        city: String,
        state: String,
        zipcode: String
    },
    beds: { type: Number, required: true },
    baths: { type: Number, required: true },
    squareFeet: { type: Number, required: true },
    amenities: [{ type: String }],
    rates: {
        nightly: Number,
        weekly: Number,
        monthly: Number
    },
    sellerInfo: {
        name: String,
        email: String,
        phone: String
    },
    imagesData: {
        type: [ImageSchema],
        required: true,
        validate: {
            validator: (value: PropertyImageData[]) => Array.isArray(value) && value.length > 0,
            message: 'At least one image is required in imagesData.',
        },
    },
    isFeatured: { type: Boolean, default: false }
}, {
    timestamps: true
});

const Property = models.Property || model<PropertyDocument>('Property', PropertySchema);
export default Property;