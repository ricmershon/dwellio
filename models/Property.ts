import { Document, Schema, Types, model, models } from 'mongoose';
import { UserInterface } from './User';

export interface PropertyInterface extends Document {
    _id: Types.ObjectId;
    owner: UserInterface;
    name: string;
    type: string;
    description?: string;
    location: {
        street: string;
        city: string;
        state: string;
        zipcode: string
    };
    beds: number;
    baths: number;
    square_feet: number;
    amenities?: Array<string>;
    rates: {
        nightly?: number;
        weekly?: number;
        monthly?: number
    };
    seller_info: {
        name: string;
        email: string;
        phone: string
    };
    images: Array<string>;
    is_featured: boolean;
    createdAt: Date;
    updatedAt: Date;
}

const PropertySchema = new Schema({
    _id: Schema.Types.ObjectId,
    owner: { type: Schema.Types.ObjectId, required: true, ref: 'User' },
    name: { type: String, required: true },
    type: { type: String, required: true },
    description: { type: String  },
    location: {
        street: String,
        city: String,
        state: String,
        zipcode: String
    },
    beds: { type: Number, required: true },
    baths: { type: Number, required: true },
    square_feet: { type: Number, required: true },
    amenities: [{ type: String }],
    rates: {
        nightly: Number,
        weekly: Number,
        monthly: Number
    },
    seller_info: {
        name: String,
        email: String,
        phone: String
    },
    images: [{ type: String }],
    is_featured: { type: Boolean, default: false }
}, {
    timestamps: true
});

// export const Property = models.Property || model('Property', PropertySchema);

export const Property = models.Property || model<PropertyInterface>('Property', PropertySchema);