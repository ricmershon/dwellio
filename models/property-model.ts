import { Document, Schema, Types, model, models } from "mongoose";

import { PropertyImageData } from "@/types/types";
import { PropertyInputType } from "@/schemas/property-schema";

export interface PropertyDocument extends Omit<PropertyInputType, "imagesData">, Document {
    owner: Types.ObjectId;
    isFeatured?: boolean;
    imagesData?: PropertyImageData[];
    createdAt: Date;
    updatedAt: Date;
}

const ImageSchema = new Schema({
    secureUrl: { type: String, required: true },
    publicId: { type: String, required: true },
    height: { type: Number, required: true },
    width: { type: Number, required: true },
}, { _id: false });

const PropertySchema = new Schema<PropertyDocument>({
    owner: { type: Schema.Types.ObjectId, required: true, ref: "User" },
    name: { type: String, required: true },
    type: { type: String, required: true },
    description: { type: String  },
    location: {
        street: String,
        city: String,
        state: String,
        zipcode: String
    },
    beds: { type: Number, required: true },
    baths: { type: Number, required: true },
    squareFeet: { type: Number, required: true },
    amenities: [{ type: String }],
    rates: {
        nightly: Number,
        weekly: Number,
        monthly: Number
    },
    sellerInfo: {
        name: String,
        email: String,
        phone: String
    },
    imagesData: {
        type: [ImageSchema],
        required: true,
        validate: {
            validator: (value: PropertyImageData[]) => Array.isArray(value) && value.length > 2,
            message: "At least three images are required.",
        },
    },
    isFeatured: { type: Boolean, default: false }
}, {
    timestamps: true
});

const Property = models.Property || model<PropertyDocument>("Property", PropertySchema);
export default Property;