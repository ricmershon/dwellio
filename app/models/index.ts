import Property, { PropertyInterface } from "@/app/models/property-model";
import User, { UserInterface} from "@/app/models/user-model";

export { Property, User };
export type { PropertyInterface, UserInterface }