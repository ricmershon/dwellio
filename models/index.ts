import Property, { PropertyDocument } from "@/models/property-model";
import User, { UserDocument} from "@/models/user-model";
import Message, { MessageDocument } from "@/models/message-model";
import StaticInputs, { StaticInputsDocument } from "@/models/static-inputs-model";

export { Property, User, Message, StaticInputs };
export type { PropertyDocument, UserDocument, MessageDocument, StaticInputsDocument }