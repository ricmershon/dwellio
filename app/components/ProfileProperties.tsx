'use client';

import { useState } from "react";
import Image from "next/image";
import Link from "next/link";
import { toast } from "react-toastify";

import { deleteProperty } from "@/app/lib/actions";
import { PropertyInterfaceWithId } from "@/app/models/property-model";

const ProfileProperties = (
    { fetchProperties }: { fetchProperties: Array<PropertyInterfaceWithId> }
) => {
    const [properties, setProperties] = useState(fetchProperties);

    const handleDeleteProperty = async (propertyId: string) => {
        const confirmed = window.confirm('Are you sure you want to delete the property?');
        if (!confirmed) {
            return
        }

        await deleteProperty(propertyId);

        const updatedProperties = properties.filter(
            (property) => property._id.toString() !== propertyId
        );
        setProperties(updatedProperties);

        toast.success('Property successfully deleted.')
    }

    return (
        <>
            {properties.length !== 0 && properties.map((property) => {
                const { street, city, state } = property.location;
                const propertyId = property._id.toString();
                return (
                    <div key={propertyId} className="mb-10">
                        <Link href={`/properties/${propertyId}`}>
                            <Image
                                className="h-32 w-full rounded-md object-cover"
                                src={property.images[0]}
                                alt="Property 1"
                                width={400}
                                height={400}
                            />
                        </Link>
                        <div className="mt-2">
                            <p className="text-lg font-semibold">{property.name}</p>
                            <p className="text-gray-600">Address: {street} {city} {state}</p>
                        </div>
                        <div className="mt-2">
                            <Link
                                href={`/properties/${propertyId}/edit`}
                                className="bg-blue-500 text-white px-3 py-3 rounded-md mr-2 hover:bg-blue-600"
                            >
                                Edit
                            </Link>
                            <button
                                className="bg-red-500 text-white px-3 py-2 rounded-md hover:bg-red-600 cursor-pointer"
                                type="button"
                                onClick={() => handleDeleteProperty(propertyId)}
                            >
                                Delete
                            </button>
                        </div>
                    </div>
                )
            })}
        </>
    );
}
 
export default ProfileProperties;