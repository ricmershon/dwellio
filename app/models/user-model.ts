import { Document, Schema, Types, model, models } from 'mongoose';

import { PropertyDocument } from '@/app/models/property-document';

export interface UserDocument extends Document {
    username: string;
    email: string;
    image: string;
    favorites: Types.ObjectId[] & PropertyDocument[]
}

const UserSchema = new Schema({
    username: { type: String, required: [true, 'Username is required'], unique: true },
    email: {
        type: String,
        unique: [true, 'Email already exists'],
        required: [true, 'Email is required']
    },
    image: { type: String },
    favorites: [{ type: Schema.Types.ObjectId, ref: 'Property' }]
}, {
    timestamps: true
});

const User = models.User || model<UserDocument>('User', UserSchema);
export default User;