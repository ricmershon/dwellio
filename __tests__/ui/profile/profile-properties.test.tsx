import { render, screen } from '@testing-library/react';
import ProfileProperties from '@/ui/profile/profile-properties';
import { PropertyDocument } from '@/models';

// Mock Next.js components
jest.mock('next/image', () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return function MockImage({ src, alt, className, ...props }: any) {
        return <img src={src} alt={alt} className={className} {...props} />;
    };
});

jest.mock('next/link', () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return function MockLink({ href, children, className }: any) {
        return <a href={href} className={className}>{children}</a>;
    };
});

// Mock DeletePropertyButton component
jest.mock('@/ui/profile/delete-property-button', () => {
    return function MockDeletePropertyButton({ propertyId }: { propertyId: string }) {
        return <button data-testid={`delete-button-${propertyId}`}>Delete</button>;
    };
});

describe('ProfileProperties Component', () => {
    const mockProperty: PropertyDocument = {
        _id: 'property123',
        name: 'Beautiful Test Property',
        type: 'House',
        description: 'A wonderful property for testing',
        location: {
            street: '123 Test St',
            city: 'Test City', 
            state: 'TX',
            zipcode: '12345'
        },
        beds: 3,
        baths: 2,
        squareFeet: 1500,
        amenities: ['WiFi', 'Pool'],
        rates: {
            nightly: 200,
            weekly: 1200,
            monthly: 4000
        },
        sellerInfo: {
            name: 'Test Seller',
            email: 'seller@test.com',
            phone: '555-1234'
        },
        owner: 'user123' as unknown as PropertyDocument['owner'],
        imagesData: [
            {
                secureUrl: 'https://test.com/image1.jpg',
                publicId: 'img1',
                width: 800,
                height: 600
            }
        ],
        isFeatured: false,
        createdAt: new Date('2024-01-01'),
        updatedAt: new Date('2024-01-01')
    } as PropertyDocument;

    const mockProperties = [
        mockProperty,
        {
            ...mockProperty,
            _id: 'property456',
            name: 'Another Test Property',
            location: {
                street: '456 Another St',
                city: 'Another City',
                state: 'CA',
                zipcode: '67890'
            },
            imagesData: [
                {
                    secureUrl: 'https://test.com/image2.jpg',
                    publicId: 'img2', 
                    width: 800,
                    height: 600
                }
            ]
        }
    ] as PropertyDocument[];

    describe('Grid Layout and Responsiveness', () => {
        it('should render responsive grid with correct breakpoint classes', () => {
            const { container } = render(<ProfileProperties properties={mockProperties} />);
            
            const gridElement = container.querySelector('.grid');
            expect(gridElement).toHaveClass(
                'grid',
                'grid-cols-2',
                'sm:grid-cols-3',
                'lg:grid-cols-4',
                'xl:grid-cols-5'
            );
        });

        it('should apply proper gap spacing', () => {
            const { container } = render(<ProfileProperties properties={mockProperties} />);
            
            const gridElement = container.querySelector('.grid');
            expect(gridElement).toHaveClass('gap-3', 'md:gap-4');
        });

        it('should maintain grid structure with single property', () => {
            const { container } = render(<ProfileProperties properties={[mockProperty]} />);
            
            const gridElement = container.querySelector('.grid');
            expect(gridElement).toBeInTheDocument();
            expect(gridElement?.children).toHaveLength(1);
        });

        it('should handle multiple properties in grid layout', () => {
            const { container } = render(<ProfileProperties properties={mockProperties} />);
            
            const gridElement = container.querySelector('.grid');
            expect(gridElement?.children).toHaveLength(2);
        });
    });

    describe('Property Card Rendering', () => {
        it('should render property cards for each property in array', () => {
            render(<ProfileProperties properties={mockProperties} />);
            
            expect(screen.getByText('Beautiful Test Property')).toBeInTheDocument();
            expect(screen.getByText('Another Test Property')).toBeInTheDocument();
        });

        it('should use property._id as React key', () => {
            const { container } = render(<ProfileProperties properties={mockProperties} />);
            
            const propertyCards = container.querySelectorAll('[key]');
            expect(propertyCards).toHaveLength(0); // Keys don't appear in DOM but should be used internally
            
            // Verify unique content instead
            expect(screen.getByText('Beautiful Test Property')).toBeInTheDocument();
            expect(screen.getByText('Another Test Property')).toBeInTheDocument();
        });

        it('should apply proper card styling', () => {
            const { container } = render(<ProfileProperties properties={[mockProperty]} />);
            
            const propertyCard = container.querySelector('.mb-4');
            expect(propertyCard).toHaveClass('mb-4', 'md:mb-6', 'rounded-lg', 'shadow-lg');
        });

        it('should maintain card structure for all properties', () => {
            const { container } = render(<ProfileProperties properties={mockProperties} />);
            
            const propertyCards = container.querySelectorAll('.mb-4.md\\:mb-6.rounded-lg.shadow-lg');
            expect(propertyCards).toHaveLength(2);
        });
    });

    describe('Property Image Handling', () => {
        it('should render property image from imagesData[0].secureUrl', () => {
            render(<ProfileProperties properties={[mockProperty]} />);
            
            const image = screen.getByAltText('property123');
            expect(image).toHaveAttribute('src', 'https://test.com/image1.jpg');
        });

        it('should link image to property details page', () => {
            render(<ProfileProperties properties={[mockProperty]} />);
            
            const links = screen.getAllByRole('link');
            const imageLink = links.find(link => link.getAttribute('href') === '/properties/property123');
            expect(imageLink).toHaveAttribute('href', '/properties/property123');
        });

        it('should apply proper image styling', () => {
            render(<ProfileProperties properties={[mockProperty]} />);
            
            const image = screen.getByAltText('property123');
            expect(image).toHaveClass('h-32', 'w-full', 'rounded-t-lg', 'object-cover');
        });

        it('should handle different image URLs for multiple properties', () => {
            render(<ProfileProperties properties={mockProperties} />);
            
            const image1 = screen.getByAltText('property123');
            const image2 = screen.getByAltText('property456');
            
            expect(image1).toHaveAttribute('src', 'https://test.com/image1.jpg');
            expect(image2).toHaveAttribute('src', 'https://test.com/image2.jpg');
        });

        it('should set proper image dimensions', () => {
            render(<ProfileProperties properties={[mockProperty]} />);
            
            const image = screen.getByAltText('property123');
            expect(image).toHaveAttribute('width', '400');
            expect(image).toHaveAttribute('height', '400');
        });
    });

    describe('Property Information Display', () => {
        it('should display property name', () => {
            render(<ProfileProperties properties={[mockProperty]} />);
            
            expect(screen.getByText('Beautiful Test Property')).toBeInTheDocument();
        });

        it('should format address as "street city state"', () => {
            render(<ProfileProperties properties={[mockProperty]} />);
            
            expect(screen.getByText('123 Test St Test City TX')).toBeInTheDocument();
        });

        it('should extract location data from property.location object', () => {
            const propertyWithDifferentLocation = {
                ...mockProperty,
                _id: 'property789',
                location: {
                    street: '789 Different Ave',
                    city: 'Different City',
                    state: 'FL',
                    zipcode: '54321'
                }
            } as PropertyDocument;

            render(<ProfileProperties properties={[propertyWithDifferentLocation]} />);
            
            expect(screen.getByText('789 Different Ave Different City FL')).toBeInTheDocument();
        });

        it('should apply appropriate text styling', () => {
            render(<ProfileProperties properties={[mockProperty]} />);
            
            const propertyName = screen.getByText('Beautiful Test Property');
            const address = screen.getByText('123 Test St Test City TX');
            
            expect(propertyName.parentElement).toBeInTheDocument();
            expect(address).toHaveClass('text-gray-800', 'text-sm');
        });

        it('should handle different property names and addresses', () => {
            render(<ProfileProperties properties={mockProperties} />);
            
            expect(screen.getByText('Beautiful Test Property')).toBeInTheDocument();
            expect(screen.getByText('Another Test Property')).toBeInTheDocument();
            expect(screen.getByText('123 Test St Test City TX')).toBeInTheDocument();
            expect(screen.getByText('456 Another St Another City CA')).toBeInTheDocument();
        });
    });

    describe('Action Buttons Integration', () => {
        it('should render Edit link with correct href', () => {
            render(<ProfileProperties properties={[mockProperty]} />);
            
            const editLink = screen.getByText('Edit');
            expect(editLink).toHaveAttribute('href', '/properties/property123/edit');
        });

        it('should apply proper Edit button styling', () => {
            render(<ProfileProperties properties={[mockProperty]} />);
            
            const editLink = screen.getByText('Edit');
            expect(editLink).toHaveClass('btn', 'btn-primary', 'mr-2');
        });

        it('should render DeletePropertyButton with propertyId prop', () => {
            render(<ProfileProperties properties={[mockProperty]} />);
            
            const deleteButton = screen.getByTestId('delete-button-property123');
            expect(deleteButton).toBeInTheDocument();
        });

        it('should maintain button layout and spacing', () => {
            const { container } = render(<ProfileProperties properties={[mockProperty]} />);
            
            const buttonContainer = container.querySelector('.mt-2.flex');
            expect(buttonContainer).toBeInTheDocument();
        });

        it('should render action buttons for each property', () => {
            render(<ProfileProperties properties={mockProperties} />);
            
            expect(screen.getByTestId('delete-button-property123')).toBeInTheDocument();
            expect(screen.getByTestId('delete-button-property456')).toBeInTheDocument();
            
            const editLinks = screen.getAllByText('Edit');
            expect(editLinks).toHaveLength(2);
        });

        it('should generate correct edit URLs for multiple properties', () => {
            render(<ProfileProperties properties={mockProperties} />);
            
            const editLinks = screen.getAllByText('Edit');
            expect(editLinks[0]).toHaveAttribute('href', '/properties/property123/edit');
            expect(editLinks[1]).toHaveAttribute('href', '/properties/property456/edit');
        });
    });

    describe('Data Handling', () => {
        it('should handle PropertyDocument interface correctly', () => {
            render(<ProfileProperties properties={[mockProperty]} />);
            
            expect(screen.getByText('Beautiful Test Property')).toBeInTheDocument();
            expect(screen.getByText('123 Test St Test City TX')).toBeInTheDocument();
            expect(screen.getByAltText('property123')).toBeInTheDocument();
        });

        it('should cast _id to string for routing', () => {
            render(<ProfileProperties properties={[mockProperty]} />);
            
            const links = screen.getAllByRole('link');
            const imageLink = links.find(link => link.getAttribute('href') === '/properties/property123');
            expect(imageLink).toHaveAttribute('href', '/properties/property123');
        });

        it('should access nested location properties safely', () => {
            const propertyWithComplexLocation = {
                ...mockProperty,
                location: {
                    street: 'Complex Street Name With Numbers 123',
                    city: 'Complex City-Name',
                    state: 'NY',
                    zipcode: '10001-1234'
                }
            } as PropertyDocument;

            render(<ProfileProperties properties={[propertyWithComplexLocation]} />);
            
            expect(screen.getByText('Complex Street Name With Numbers 123 Complex City-Name NY')).toBeInTheDocument();
        });

        it('should handle arrays with different lengths', () => {
            const singleProperty = [mockProperty];
            const multipleProperties = mockProperties;
            
            const { rerender } = render(<ProfileProperties properties={singleProperty} />);
            expect(screen.getAllByText('Edit')).toHaveLength(1);
            
            rerender(<ProfileProperties properties={multipleProperties} />);
            expect(screen.getAllByText('Edit')).toHaveLength(2);
        });
    });

    describe('Empty State Handling', () => {
        it('should handle empty properties array gracefully', () => {
            const { container } = render(<ProfileProperties properties={[]} />);
            
            const gridElement = container.querySelector('.grid');
            expect(gridElement).toBeInTheDocument();
            expect(gridElement?.children).toHaveLength(0);
        });

        it('should not render any cards when properties.length === 0', () => {
            render(<ProfileProperties properties={[]} />);
            
            expect(screen.queryByText('Edit')).not.toBeInTheDocument();
            expect(screen.queryByRole('img')).not.toBeInTheDocument();
        });

        it('should maintain grid structure regardless of content', () => {
            const { container } = render(<ProfileProperties properties={[]} />);
            
            const gridElement = container.querySelector('.grid');
            expect(gridElement).toHaveClass(
                'grid',
                'grid-cols-2',
                'sm:grid-cols-3',
                'lg:grid-cols-4',
                'xl:grid-cols-5',
                'gap-3',
                'md:gap-4'
            );
        });
    });

    describe('Edge Cases and Error Handling', () => {
        it('should handle properties without imagesData gracefully', () => {
            const propertyWithoutImages = {
                ...mockProperty,
                imagesData: [{
                    secureUrl: '',
                    publicId: '',
                    width: 0,
                    height: 0
                }]
            } as unknown as PropertyDocument;

            // This should not crash but may show broken image
            expect(() => {
                render(<ProfileProperties properties={[propertyWithoutImages]} />);
            }).not.toThrow();
        });

        it('should handle properties with missing location fields', () => {
            const propertyWithIncompleteLocation = {
                ...mockProperty,
                location: {
                    street: 'Only Street',
                    city: '',
                    state: '',
                    zipcode: '12345'
                }
            } as unknown as PropertyDocument;

            render(<ProfileProperties properties={[propertyWithIncompleteLocation]} />);
            
            expect(screen.getByText('Only Street')).toBeInTheDocument();
        });

        it('should handle properties with special characters in names', () => {
            const propertyWithSpecialChars = {
                ...mockProperty,
                name: "Property with Special Chars !@#$%^&*()"
            } as PropertyDocument;

            render(<ProfileProperties properties={[propertyWithSpecialChars]} />);
            
            expect(screen.getByText("Property with Special Chars !@#$%^&*()")).toBeInTheDocument();
        });

        it('should handle very long property names', () => {
            const propertyWithLongName = {
                ...mockProperty,
                name: "This is a very long property name that should be handled gracefully by the component without breaking the layout or causing display issues"
            } as PropertyDocument;

            render(<ProfileProperties properties={[propertyWithLongName]} />);
            
            expect(screen.getByText("This is a very long property name that should be handled gracefully by the component without breaking the layout or causing display issues")).toBeInTheDocument();
        });
    });

    describe('Snapshots', () => {
        it('should match snapshot with empty properties array', () => {
            const { container } = render(<ProfileProperties properties={[]} />);
            expect(container.firstChild).toMatchSnapshot();
        });

        it('should match snapshot with single property', () => {
            const { container } = render(<ProfileProperties properties={[mockProperty]} />);
            expect(container.firstChild).toMatchSnapshot();
        });

        it('should match snapshot with multiple properties', () => {
            const { container } = render(<ProfileProperties properties={mockProperties} />);
            expect(container.firstChild).toMatchSnapshot();
        });

        it('should match snapshot with different property types', () => {
            const differentTypeProperty = {
                ...mockProperty,
                _id: 'apartment123',
                name: 'Downtown Apartment',
                type: 'Apartment',
                location: {
                    street: '456 City Ave',
                    city: 'Metro City',
                    state: 'NY',
                    zipcode: '10001'
                }
            } as PropertyDocument;

            const { container } = render(<ProfileProperties properties={[differentTypeProperty]} />);
            expect(container.firstChild).toMatchSnapshot();
        });
    });
});