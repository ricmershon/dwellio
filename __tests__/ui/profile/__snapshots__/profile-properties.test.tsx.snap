// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ProfileProperties Component Snapshots should match snapshot with different property types 1`] = `
<div
  class="grid grid-cols-2 sm:grid-cols-3 lg:grid-cols-4 xl:grid-cols-5 gap-3 md:gap-4"
>
  <div
    class="mb-4 md:mb-6 rounded-lg shadow-lg"
  >
    <a
      href="/properties/apartment123"
    >
      <img
        alt="apartment123"
        class="h-32 w-full rounded-t-lg object-cover"
        height="400"
        src="https://test.com/image1.jpg"
        width="400"
      />
    </a>
    <div
      class="p-4"
    >
      <div
        class=""
      >
        <p>
          Downtown Apartment
        </p>
        <p
          class="text-gray-800 text-sm"
        >
          456 City Ave
           
          Metro City
           
          NY
        </p>
      </div>
      <div
        class="mt-2 flex"
      >
        <a
          class="btn btn-primary mr-2"
          href="/properties/apartment123/edit"
        >
          Edit
        </a>
        <button
          data-testid="delete-button-apartment123"
        >
          Delete
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ProfileProperties Component Snapshots should match snapshot with empty properties array 1`] = `
<div
  class="grid grid-cols-2 sm:grid-cols-3 lg:grid-cols-4 xl:grid-cols-5 gap-3 md:gap-4"
/>
`;

exports[`ProfileProperties Component Snapshots should match snapshot with multiple properties 1`] = `
<div
  class="grid grid-cols-2 sm:grid-cols-3 lg:grid-cols-4 xl:grid-cols-5 gap-3 md:gap-4"
>
  <div
    class="mb-4 md:mb-6 rounded-lg shadow-lg"
  >
    <a
      href="/properties/property123"
    >
      <img
        alt="property123"
        class="h-32 w-full rounded-t-lg object-cover"
        height="400"
        src="https://test.com/image1.jpg"
        width="400"
      />
    </a>
    <div
      class="p-4"
    >
      <div
        class=""
      >
        <p>
          Beautiful Test Property
        </p>
        <p
          class="text-gray-800 text-sm"
        >
          123 Test St
           
          Test City
           
          TX
        </p>
      </div>
      <div
        class="mt-2 flex"
      >
        <a
          class="btn btn-primary mr-2"
          href="/properties/property123/edit"
        >
          Edit
        </a>
        <button
          data-testid="delete-button-property123"
        >
          Delete
        </button>
      </div>
    </div>
  </div>
  <div
    class="mb-4 md:mb-6 rounded-lg shadow-lg"
  >
    <a
      href="/properties/property456"
    >
      <img
        alt="property456"
        class="h-32 w-full rounded-t-lg object-cover"
        height="400"
        src="https://test.com/image2.jpg"
        width="400"
      />
    </a>
    <div
      class="p-4"
    >
      <div
        class=""
      >
        <p>
          Another Test Property
        </p>
        <p
          class="text-gray-800 text-sm"
        >
          456 Another St
           
          Another City
           
          CA
        </p>
      </div>
      <div
        class="mt-2 flex"
      >
        <a
          class="btn btn-primary mr-2"
          href="/properties/property456/edit"
        >
          Edit
        </a>
        <button
          data-testid="delete-button-property456"
        >
          Delete
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ProfileProperties Component Snapshots should match snapshot with single property 1`] = `
<div
  class="grid grid-cols-2 sm:grid-cols-3 lg:grid-cols-4 xl:grid-cols-5 gap-3 md:gap-4"
>
  <div
    class="mb-4 md:mb-6 rounded-lg shadow-lg"
  >
    <a
      href="/properties/property123"
    >
      <img
        alt="property123"
        class="h-32 w-full rounded-t-lg object-cover"
        height="400"
        src="https://test.com/image1.jpg"
        width="400"
      />
    </a>
    <div
      class="p-4"
    >
      <div
        class=""
      >
        <p>
          Beautiful Test Property
        </p>
        <p
          class="text-gray-800 text-sm"
        >
          123 Test St
           
          Test City
           
          TX
        </p>
      </div>
      <div
        class="mt-2 flex"
      >
        <a
          class="btn btn-primary mr-2"
          href="/properties/property123/edit"
        >
          Edit
        </a>
        <button
          data-testid="delete-button-property123"
        >
          Delete
        </button>
      </div>
    </div>
  </div>
</div>
`;
